### PR TITLE
Only add entries to SourceToDef dynmaps when they come from the same file

### PR DESCRIPTION
**Pull request by Veykril**
Fixes https://github.com/rust-analyzer/rust-analyzer/issues/9919

Running the test as described in the issue I do not get any eprintln output at all anymore.

### CODE REVIEW

File: crates/hir/src/semantics/source_to_def.rs [hunk 1]
```diff
@@ -106,7 +106,7 @@ use syntax::{
 
 use crate::{db::HirDatabase, InFile};
 
-pub(super) type SourceToDefCache = FxHashMap<ChildContainer, DynMap>;
+pub(super) type SourceToDefCache = FxHashMap<ChildContainer, FxHashMap<HirFileId, DynMap>>;
```
Comment (lnicola) [inline crates/hir/src/semantics/source_to_def.rs:109]: I'd go for the compound keys if it's not too complicated.

Comment (Veykril) [reply crates/hir/src/semantics/source_to_def.rs:109]: Oh right, I forgot to switch to that. Was more interested in seeing whether this passes the tests first.

**Comment by matklad** [comment]
Curious, do you get a perf increase out of it?

**Comment by Veykril** [comment]
I do actually:
- before: `203ms - dot completion`
- after: `69ms - dot completion`

**Comment by matklad** [comment]
this is debug, right?

**Comment by Veykril** [comment]
This was with 
```
$ cargo test --release --package rust-analyzer --lib --quiet -- integrated_benchmarks::integrated_completion --nocapture
```

**Comment by Veykril** [comment]
I can rerun to make sure this wasnt my system doing something funny though.

**Comment by matklad** [comment]
Yeah, 200 -> 60 feels strange. I get more like 60 -> 40 (still a nice bump)

File: crates/hir_def/src/child_by_source.rs [hunk 2]
```diff
@@ -17,145 +18,196 @@ use crate::{
...
+    fn child_by_source_to(&self, db: &dyn DefDatabase, file_id: HirFileId, map: &mut DynMap);
 }
 
 impl ChildBySource for TraitId {
-    fn child_by_source_to(&self, db: &dyn DefDatabase, res: &mut DynMap) {
+    fn child_by_source_to(&self, db: &dyn DefDatabase, file_id: HirFileId, res: &mut DynMap) {
         let data = db.trait_data(*self);
-        for (_name, item) in data.items.iter() {
-            match *item {
+        for &(_, item) in data.items.iter() {
```
Comment (matklad) [inline crates/hir_def/src/child_by_source.rs:32]: the `_name` is intentional -- it's useful to see what you are ignoring.

File: crates/hir_def/src/child_by_source.rs [hunk 3]
```diff
@@ -17,145 +18,196 @@ use crate::{
...
+        self.child_by_source_to(db, file_id, &mut res);
         res
     }
-    fn child_by_source_to(&self, db: &dyn DefDatabase, map: &mut DynMap);
+    fn child_by_source_to(&self, db: &dyn DefDatabase, file_id: HirFileId, map: &mut DynMap);
 }
 
 impl ChildBySource for TraitId {
-    fn child_by_source_to(&self, db: &dyn DefDatabase, res: &mut DynMap) {
+    fn child_by_source_to(&self, db: &dyn DefDatabase, file_id: HirFileId, res: &mut DynMap) {
```
Comment (matklad) [inline crates/hir_def/src/child_by_source.rs:30]: I'd probably swap the order of file_id and res, to keep accumulator/context on the left, as per style guide.

File: crates/hir_def/src/child_by_source.rs [hunk 4]
```diff
@@ -17,145 +18,196 @@ use crate::{
...
+                    }
                 }
                 AssocItemId::TypeAliasId(ty) => {
-                    let src = ty.lookup(db).source(db);
-                    res[keys::TYPE_ALIAS].insert(src, ty)
+                    let loc = ty.lookup(db);
+                    if loc.id.file_id() == file_id {
+                        let src = loc.source(db);
+                        res[keys::TYPE_ALIAS].insert(src, ty)
+                    }
```
Comment (matklad) [inline crates/hir_def/src/child_by_source.rs:53]: I am afraid of the day we implement [code duplicates search](https://www.jetbrains.com/help/idea/find-and-replace-code-duplicates.html)

**Comment by Veykril** [comment]
> Yeah, 200 -> 60 feels strange.
Just rerun both versions again with the same result 🤔

**Pull request merged**